pkg/protocol/protocol_manager.go

```diff
@@ -0,0 +1,231 @@
+		}
+	}
+	return false
+}
+
+// NewManager creates a new Manager.
+func NewManager(storage *storage.Storage, initProtoParas *iotago.ProtocolParameters) *Manager {
+	return &Manager{
+		Events: &Events{
+			NextMilestoneUnsupported: events.NewEvent(protoParasMsOptCaller),
+			CriticalErrors:           events.NewEvent(events.ErrorCaller),
+		},
+		storage: storage,
+		current: initProtoParas,
+		pending: nil,
+	}
+}
+
+func protoParasMsOptCaller(handler interface{}, params ...interface{}) {
+	handler.(func(protoParas *iotago.ProtocolParamsMilestoneOpt))(params[0].(*iotago.ProtocolParamsMilestoneOpt))
+}
+
+// Events are events happening around the Manager.
+type Events struct {
+	// Emits protocol parameters for the unsupported milestone one milestone before.
+	NextMilestoneUnsupported *events.Event
+	// Emits critical errors.
+	CriticalErrors *events.Event
+}
+
+// Manager handles the knowledge about current, pending and supported protocol versions and parameters.
+type Manager struct {
+	// Events holds the events happening within the Manager.
+	Events      *Events
+	storage     *storage.Storage
+	syncManager *syncmanager.SyncManager
+	currentRWMu sync.RWMutex
+	current     *iotago.ProtocolParameters
+	pendingRWMu sync.RWMutex
+	pending     []*iotago.ProtocolParamsMilestoneOpt
+}
+
+// Init initialises the Manager by loading the last stored or persisting the parameters passed in via constructor.
+// If the Manager got initialised with no storage, then Manager.Init() is a no-op and initProtoParas are used instead.
+func (m *Manager) Init() error {
+	// can only run with provided protocol parameters
+	if m.storage == nil {
+		return nil
+	}
+
+	m.currentRWMu.Lock()
+	defer m.currentRWMu.Unlock()
+
+	currentProtoParas, err := m.storage.ProtocolParameters()
+	if err != nil {
+		return err
+	}
+
+	// aka store therefore what the manager was initialised with
+	if currentProtoParas == nil {
+		if err := m.storage.StoreProtocolParameters(m.current); err != nil {
+			return fmt.Errorf("unable to persist init protocol parameters: %w", err)
+		}
+		return nil
+	}
+
+	m.current = currentProtoParas
+
+	return nil
+}
+
+// LoadPending examines the database back to below max depth for pending protocol parameter changes.
+func (m *Manager) LoadPending(syncManager *syncmanager.SyncManager) {
+	m.pendingRWMu.Lock()
+	defer m.pendingRWMu.Unlock()
+
+	// examine below max depth milestone to reconstruct pending protocol changes
+	confMsIndex := syncManager.ConfirmedMilestoneIndex()
+	belowMaxDepthTarget := confMsIndex - milestone.Index(m.current.BelowMaxDepth)
+	for i := belowMaxDepthTarget; i < confMsIndex; i-- {
+		if protoParasMsOpt := m.readProtocolParasFromMilestone(i); protoParasMsOpt != nil {
+			m.pending = append(m.pending, protoParasMsOpt)
+		}
+	}
+}
+
+func (m *Manager) readProtocolParasFromMilestone(index milestone.Index) *iotago.ProtocolParamsMilestoneOpt {
+	cachedMs := m.storage.CachedMilestoneByIndexOrNil(index)
+	defer cachedMs.Release(true)
```

Comment:
You can not call Release on it, in case it is nil

Reply:
fixed